vaporetto/src/char_scorer/boundary_scorer.rs

```diff
@@ -0,0 +1,112 @@
+    fn borrow_decode<D: BorrowDecoder<'de>>(decoder: &mut D) -> Result<Self, DecodeError> {
+        let pma_data: &[u8] = BorrowDecode::borrow_decode(decoder)?;
+        #[cfg(not(feature = "charwise-pma"))]
+        let (pma, _) =
+            unsafe { DoubleArrayAhoCorasick::deserialize_from_slice_unchecked(pma_data) };
+        #[cfg(feature = "charwise-pma")]
+        let (pma, _) =
+            unsafe { CharwiseDoubleArrayAhoCorasick::deserialize_from_slice_unchecked(pma_data) };
+        Ok(Self {
+            pma,
+            weights: Decode::decode(decoder)?,
+        })
+    }
+}
+
+impl Encode for CharScorerBoundary {
+    fn encode<E: Encoder>(&self, encoder: &mut E) -> Result<(), EncodeError> {
+        let pma_data = self.pma.serialize_to_vec();
+        Encode::encode(&pma_data, encoder)?;
+        Encode::encode(&self.weights, encoder)?;
+        Ok(())
+    }
+}
+
+impl CharScorerBoundary {
+    pub fn new(
+        ngram_model: NgramModel<String>,
+        dict_model: DictModel,
+        window_size: u8,
+    ) -> Result<Self> {
+        let mut merger = CharWeightMerger::default();
+        for d in ngram_model.0 {
+            let weight = PositionalWeight::new(-i16::from(window_size), d.weights);
+            merger.add(d.ngram, weight);
+        }
+        for d in dict_model.0 {
+            let word_len = d.word.chars().count();
+            let word_len = i16::try_from(word_len).map_err(|_| {
+                VaporettoError::invalid_model(
+                    "words must be shorter than or equal to 32767 characters",
+                )
+            })?;
+            let weight = PositionalWeight::new(-word_len, d.weights);
+            merger.add(d.word, weight);
+        }
+        let mut ngrams = vec![];
+        let mut weights = vec![];
+        for (ngram, weight) in merger.merge() {
+            ngrams.push(ngram);
+            weights.push(weight.into());
+        }
+        #[cfg(not(feature = "charwise-pma"))]
+        let pma = DoubleArrayAhoCorasick::new(ngrams)
+            .map_err(|_| VaporettoError::invalid_model("failed to build the automaton"))?;
+        #[cfg(feature = "charwise-pma")]
+        let pma = CharwiseDoubleArrayAhoCorasick::new(ngrams)
+            .map_err(|_| VaporettoError::invalid_model("failed to build the automaton"))?;
+        Ok(Self { pma, weights })
+    }
+
+    #[allow(clippy::cast_possible_wrap)]
+    #[inline(always)]
+    pub fn add_scores<'a, 'b>(&self, sentence: &mut Sentence<'a, 'b>) {
+        sentence.char_pma_states.clear();
+        #[cfg(not(feature = "charwise-pma"))]
+        let it = self
+            .pma
+            .find_overlapping_no_suffix_iter(sentence.text.as_bytes());
+        #[cfg(feature = "charwise-pma")]
+        let it = self.pma.find_overlapping_no_suffix_iter(&sentence.text);
+        for m in it {
+            let end = unsafe { sentence.str_to_char_pos(m.end()) };
+            let weight = unsafe { self.weights.get_unchecked(m.value()) };
```

Comment:
```suggestion
            debug_assert!(m.end() <= sentence.text.len());
            let end = unsafe { sentence.str_to_char_pos(m.end()) };
            debug_assert!(m.end() == 0 || end != 0);
            debug_assert!(m.value() <= self.weights.len());
            let weight = unsafe { self.weights.get_unchecked(m.value()) };
```
These unsafe blocks are implemented by trusting that the third-party Daachorse works correctly.
So, debug assertions would be needed at least.

Reply:
Thanks.
I changed to the following two conditions. Is it correct?

```rust
debug_assert!(m.end() != 0 && m.end() <= sentence.text.len());
debug_assert!(m.value() < self.weights.len());
```

Reply:
@vbkaisetsu Almost yes. In addition, the following line is needed
```rust
let end = unsafe { sentence.str_to_char_pos(m.end()) };
debug_assert!(end != 0);
```
because `end == 0` means that `m.end()` indicates an invalid position in UTF8 format when `m.end() != 0`.

Reply:
@kampersanda 
I see.
But I think this assertion is redundant because this code is not a test for daachorse.
Vaporetto should trust the value returned from daachorse.

Reply:
The library is responsible for returning correct values. It is not the responsibility of the caller.

Reply:
@vbkaisetsu I understand. I agree with keeping debug assertions only for our unsafe code blocks.